### PR TITLE
feat: Add JsonExtractor for structured JSON file parsing in RAG

### LINKED ISSUE #24529: Add JsonExtractor for structured JSON file parsing in RAG
### Self Checks

- [x] I have read the [Contributing Guide](https://github.com/langgenius/dify/blob/main/CONTRIBUTING.md) and [Language Policy](https://github.com/langgenius/dify/issues/1542).
- [x] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [x] I confirm that I am using English to submit this report, otherwise it will be closed.
- [x] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing? Tell me about your story.

# Feature Request: Add JSON File Support to RAG Document Extractor

## Feature Description

Add native support for JSON file extraction in Dify's RAG document processing system. Currently, the RAG extractor supports various document formats like PDF, DOCX, TXT, etc., but lacks support for JSON files, which are commonly used for structured data storage and API responses.

## Use Case

**Primary Use Cases:**
- **API Documentation**: Extract content from JSON-formatted API documentation and responses
- **Configuration Files**: Process JSON configuration files for knowledge base creation
- **Data Export**: Handle JSON exports from databases, CRM systems, or other applications
- **Structured Data**: Process structured data stored in JSON format for RAG applications
- **Log Files**: Extract information from JSON-formatted log files

**Example Scenario:**
"I was trying to use JSON files containing product catalogs for my RAG application, and I expected the system to extract the structured data and make it searchable. However, the current extractor doesn't support JSON files, so I had to manually convert them to text format, which loses the structural information and makes the data less useful for retrieval."

### Core Functionality
- **JSON Parsing**: Parse valid JSON files and extract text content from string values
- **Nested Structure Handling**: Support nested JSON objects and arrays
- **Key-Value Extraction**: Extract both keys and values as meaningful content
- **Array Processing**: Handle JSON arrays with proper indexing
- **Metadata Preservation**: Maintain JSON structure information in document metadata

### Technical Specifications
- **File Extensions**: Support `.json` and `.jsonl` (JSON Lines) formats
- **Encoding Support**: Handle UTF-8, UTF-16, and other common encodings
- **Size Limits**: Implement reasonable file size limits to prevent memory issues
- **Error Handling**: Graceful handling of malformed JSON files
- **Content Flattening**: Convert nested JSON structures into searchable text chunks

### Integration Requirements
- **ExtractProcessor Integration**: Seamlessly integrate with existing `ExtractProcessor` class
- **Document Chunking**: Support proper text chunking for large JSON files
- **Metadata Enhancement**: Add JSON-specific metadata (structure depth, key paths, data types)
- **Testing Coverage**: Comprehensive unit tests following existing patterns

## Expected Behavior

1. **File Detection**: Automatically detect JSON files by extension
2. **Content Extraction**: Extract meaningful text content from JSON structures
3. **Structure Preservation**: Maintain hierarchical information in metadata
4. **Chunk Generation**: Create appropriate text chunks for RAG processing
5. **Error Recovery**: Handle parsing errors gracefully with informative messages

## Additional Context

- This feature would complement the existing document extractors and follow the same architectural patterns
- JSON support is essential for modern applications that work with API data and structured content
- The implementation should be consistent with other extractors in terms of error handling and metadata structure
- Consider supporting both pretty-printed and minified JSON formats

## Priority

Medium Priority - This feature would significantly enhance Dify's ability to process structured data sources commonly used in modern applications.

## Related Issues

- Related to overall document extraction improvements
- Part of expanding supported file formats for RAG applications

### 2. Additional context or comments

_No response_

### 3. Can you help us with this feature?

- [x] I am interested in contributing to this feature.

### PR DESCRIPTION
**Pull request by brianxiadong**
fixes #24529  

This PR introduces a new JsonExtractor class that enables structured parsing of JSON files in the RAG (Retrieval-Augmented Generation) system. Previously, JSON files were processed as plain text by the TextExtractor , which resulted in loss of structural information and poor retrieval performance.

Key Features:

- Structured JSON Parsing : Converts JSON objects and arrays into individual documents while preserving key-value relationships
- Nested Structure Support : Handles complex nested JSON structures with proper flattening
- Rich Metadata : Includes source path, object index, type information, key lists, and searchable text
- Error Handling : Robust error handling for malformed JSON files
- Comprehensive Testing : Full test coverage including edge cases and error scenarios
Files Changed:

- ✅ Added : api/core/rag/extractor/json_extractor.py - Main JsonExtractor implementation
- ✅ Added : api/tests/unit_tests/core/rag/extractor/test_json_extractor.py - Comprehensive test suite
- ✅ Modified : api/core/rag/extractor/extract_processor.py - Added .json extension support
Benefits:

- Improved Retrieval : Better semantic understanding of JSON content for RAG applications
- Preserved Structure : Maintains JSON hierarchy and relationships in extracted documents
- Enhanced Metadata : Rich metadata enables more precise document filtering and retrieval
- Better User Experience : More accurate responses when working with JSON data sources
Testing:

- ✅ Unit tests cover JSON objects, arrays, nested structures, and error cases
- ✅ All existing tests pass
- ✅ Code follows project conventions and includes proper English documentation
Technical Implementation:

- Follows existing extractor patterns (similar to CSVExtractor)
- Uses Python's built-in json module for parsing
- Implements proper encoding detection and error handling
- Generates individual Document objects for each top-level JSON element
- Includes comprehensive metadata for enhanced retrieval capabilities

### CODE REVIEW

File: api/core/rag/extractor/json_extractor.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        Returns:
+            Document object with formatted content and metadata
+        """
+        # Convert object to formatted JSON string
+        content = json.dumps(obj, ensure_ascii=False, indent=2)
+
+        # Determine object type and extract key information
+        obj_type = type(obj).__name__
+        metadata = {"source": self._file_path, "index": index, "type": obj_type}
+
```
[the remark this answers is not in the document]
Comment (brianxiadong) [reply api/core/rag/extractor/json_extractor.py:104]: resolved
